File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/codeclass/JooqCodeClassRepoTest.java
package ch.difty.scipamato.core.persistence.codeclass;

import static ch.difty.scipamato.common.TestUtils.assertDegenerateSupplierParameter;

import org.jooq.DSLContext;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import ch.difty.scipamato.common.DateTimeService;

@RunWith(MockitoJUnitRunner.class)
public class JooqCodeClassRepoTest {

    @Mock
    private DSLContext dslContextMock;

    @Mock
    private DateTimeService dtsMock;

    private JooqCodeClassRepo repo;

    @Before
    public void setUp() {
        repo = new JooqCodeClassRepo(dslContextMock, dtsMock);
    }

    @Test
    public void degenerateConstruction_withNullDsl_throws() {
        assertDegenerateSupplierParameter(() -> new JooqCodeClassRepo(null, dtsMock), "dsl");
    }

    @Test
    public void degenerateConstruction_withNullDateTimeService_throws() {
        assertDegenerateSupplierParameter(() -> new JooqCodeClassRepo(dslContextMock, null), "dateTimeService");
    }

    @Test
    public void finding_withNullLanguageId_throws() {
        assertDegenerateSupplierParameter(() -> repo.find(null), "languageCode");
    }

    @Test
    public void insertingOrUpdating_withNullCodeClassDefinition_throws() {
        assertDegenerateSupplierParameter(() -> repo.saveOrUpdate(null), "codeClassDefinition");
    }

    @Test
    public void findingCodeClassDefinition_withNullId_throws() {
        assertDegenerateSupplierParameter(() -> repo.findCodeClassDefinition(null), "id");
    }

    @Test
    public void deleting_withNullId_throws() {
        assertDegenerateSupplierParameter(() -> repo.delete(null, 1), "id");
    }

}

File: core/core-persistence-jooq/src/test/kotlin/ch/difty/scipamato/core/persistence/codeclass/JooqCodeClassRepoTest.kt
package ch.difty.scipamato.core.persistence.codeclass

import ch.difty.scipamato.common.DateTimeService
import ch.difty.scipamato.core.db.tables.CodeClassTr.CODE_CLASS_TR
import ch.difty.scipamato.core.db.tables.records.CodeClassTrRecord
import ch.difty.scipamato.core.entity.codeclass.CodeClassTranslation
import ch.difty.scipamato.core.persistence.OptimisticLockingException
import io.mockk.confirmVerified
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.amshove.kluent.invoking
import org.amshove.kluent.shouldThrow
import org.amshove.kluent.withMessage
import org.jooq.DSLContext
import org.jooq.Result
import org.junit.jupiter.api.Test

internal class JooqCodeClassRepoTest {

    private val dslContextMock = mockk<DSLContext>()
    private val dtsMock = mockk<DateTimeService>()
    private var repo = JooqCodeClassRepo(dslContextMock, dtsMock)

    @Suppress("SpellCheckingInspection")
    @Test
    fun removingObsoletePersistedRecords() {
        val cct = CodeClassTranslation(1, "de", "cc1", "", 1)
        val cctr1 = mockk<CodeClassTrRecord>(relaxed = true) {
            every { get(CODE_CLASS_TR.ID) } returns 1
        }
        val cctr2 = mockk<CodeClassTrRecord>(relaxed = true) {
            every { get(CODE_CLASS_TR.ID) } returns 2
        }
        val resultMock: Result<CodeClassTrRecord> = mockk {
            every { iterator() } returns mockk {
                every { hasNext() } returnsMany listOf(true, true, false)
                every { next() } returnsMany listOf(cctr1, cctr2)
            }
        }

        repo.removeObsoletePersistedRecordsFor(resultMock, listOf(cct))

        verify { resultMock.iterator() }
        verify { cctr1.get(CODE_CLASS_TR.ID) }
        verify { cctr2.get(CODE_CLASS_TR.ID) }
        verify { cctr2.delete() }

        confirmVerified(resultMock, cctr1, cctr2)
    }

    @Test
    fun consideringAdding_withNullRecord_throwsOptimisticLockingException() {
        invoking {
            repo.considerAdding(null, ArrayList(), CodeClassTranslation(1, "de", "c1", "comm", 10))
        } shouldThrow OptimisticLockingException::class withMessage
            "Record in table 'code_class_tr' has been modified prior to the update attempt. " +
            "Aborting.... [CodeClassTranslation(description=comm)]"
    }

    @Test
    fun logOrThrow_withDeleteCount0_throws() {
        invoking {
            repo.logOrThrow(0, 1, "deletedObject")
        } shouldThrow OptimisticLockingException::class withMessage
            "Record in table 'code_class' has been modified prior to the delete attempt. " +
            "Aborting.... [deletedObject]"
    }
}
